### PR TITLE
[SPIR-V] Cherry-pick of "Add SPIR-V 1.4 checks"

### PR DESCRIPTION
**Pull request by MrSidims**
One of the patches missing in intel/llvm

Signed-off-by: Sidorov, Dmitry dmitry.sidorov@intel.com

Original commit:
KhronosGroup/SPIRV-LLVM-Translator@c5b3c8e3283b

Second attempt

### CODE REVIEW

**Comment by MrSidims** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1414

**Comment by MrSidims** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1414

**Comment by MrSidims** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1477

**Comment by MrSidims** [comment]
/verify with https://github.com/intel/llvm-test-suite/pull/1479

**Comment by MrSidims** [comment]
select_device.cpp is unrelated test failure, reported internally
level_zero_batch_barrier.cpp is unrelated test failure, that is being fixed in https://github.com/intel/llvm-test-suite/pull/1481

 SubGroupMask/Basic.cpp on windows to investigate

File: llvm-spirv/lib/SPIRV/SPIRVWriter.cpp [hunk 1]
```diff
@@ -2585,10 +2588,10 @@ bool LLVMToSPIRVBase::transDecoration(Value *V, SPIRVValue *BV) {
 
   if (auto BVO = dyn_cast_or_null<OverflowingBinaryOperator>(V)) {
     if (BVO->hasNoSignedWrap()) {
-      BV->setNoSignedWrap(true);
+      BV->setNoIntegerDecorationWrap<DecorationNoSignedWrap>(true);
```
Comment (asudarsa) [inline llvm-spirv/lib/SPIRV/SPIRVWriter.cpp:2591]: Are these changes relevant to the switch to 1.4?

File: llvm-spirv/lib/SPIRV/SPIRVWriter.cpp [hunk 2]
```diff
@@ -4883,43 +4886,34 @@ bool LLVMToSPIRVBase::transExecutionMode() {
...
       } break;
       case spv::ExecutionModeNoGlobalOffsetINTEL: {
-        if (BM->isAllowedToUseExtension(
-                ExtensionID::SPV_INTEL_kernel_attributes)) {
-          BF->addExecutionMode(BM->add(
-              new SPIRVExecutionMode(BF, static_cast<ExecutionMode>(EMode))));
-          BM->addExtension(ExtensionID::SPV_INTEL_kernel_attributes);
-          BM->addCapability(CapabilityKernelAttributesINTEL);
-        }
+        if (!BM->isAllowedToUseExtension(
```
Comment (asudarsa) [inline llvm-spirv/lib/SPIRV/SPIRVWriter.cpp:4889]: Why is this change required? Nesting level does not look too bad.

File: llvm-spirv/lib/SPIRV/SPIRVWriter.cpp [hunk 3]
```diff
@@ -4883,43 +4886,34 @@ bool LLVMToSPIRVBase::transExecutionMode() {
...
       case spv::ExecutionModeVecTypeHint:
       case spv::ExecutionModeSubgroupSize:
-      case spv::ExecutionModeSubgroupsPerWorkgroup: {
-        unsigned X;
-        N.get(X);
-        BF->addExecutionMode(BM->add(
-            new SPIRVExecutionMode(BF, static_cast<ExecutionMode>(EMode), X)));
-      } break;
+      case spv::ExecutionModeSubgroupsPerWorkgroup:
+        AddSingleArgExecutionMode(static_cast<ExecutionMode>(EMode));
```
Comment (asudarsa) [inline llvm-spirv/lib/SPIRV/SPIRVWriter.cpp:4900]: Same comment as above. Does this fit into the scope of this PR? Thanks

File: llvm-spirv/lib/SPIRV/libSPIRV/SPIRVValue.cpp [hunk 4]
```diff
@@ -75,45 +75,10 @@ bool SPIRVValue::hasNoSignedWrap() const {
...
-  } else {
-    SPIRVDBG(spvdbgs() << "Skip setting nsw for obj " << Id << "\n")
-  }
-}
-
 bool SPIRVValue::hasNoUnsignedWrap() const {
   return hasDecorate(DecorationNoUnsignedWrap);
 }
 
-void SPIRVValue::setNoUnsignedWrap(bool HasNoUnsignedWrap) {
```
Comment (asudarsa) [inline llvm-spirv/lib/SPIRV/libSPIRV/SPIRVValue.cpp:99]: I suppose these need to be removed as we move to 1.4?

**Review by asudarsa — APPROVED** [review]
LGTM overall. Just some nitpicks on some changes. But I suppose I am also not looking at the full commit here. Ok to merge. Thanks. And yes, sorry for delay in review.

File: llvm-spirv/lib/SPIRV/libSPIRV/SPIRVValue.cpp (hunk 4 above)
Comment (MrSidims) [reply llvm-spirv/lib/SPIRV/libSPIRV/SPIRVValue.cpp:99]: Not sure if I understand the comment. The code is removed, isn't it?

File: llvm-spirv/lib/SPIRV/SPIRVWriter.cpp (hunk 2 above)
Comment (MrSidims) [reply llvm-spirv/lib/SPIRV/SPIRVWriter.cpp:4889]: The PR is just a cherry-pick of https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/1177 .

There I made this change looking through the code back then (now I prefer not to mix up clean up changes within non-NFC patch) for two reasons:
1. LLVM code style recommends to do nesting like this if possible;
2. It brings minor (1 or 2 clocks) performance gains, since, I believe in general cases when SPV_INTEL_kernel_attributes is not supported more than those where it is supported.

File: llvm-spirv/lib/SPIRV/SPIRVWriter.cpp (hunk 3 above)
Comment (MrSidims) [reply llvm-spirv/lib/SPIRV/SPIRVWriter.cpp:4900]: Same as above. But I do believe, that @romanovvlad would be happy to stop seeing 'unused variable' warning during SYCL runtime build.

**Review by bader — CHANGES_REQUESTED** [review]
Please, resolve merge conflicts.

**Comment by MrSidims** [comment]
>> InvalidModule: Invalid SPIR-V module: incorrect SPIR-V version number 1.4 (66560) - it conflicts with --spirv-max-version which is set to 1.2 (66048)

Seems like GPU SPIR-V consumer additionally restricts max version of supported SPIR-V module and it's not yet fixed.

**Comment by MrSidims** [comment]
Other issues are found - seems like CPU and GPU SPIR-V consumers restrict max to 1.2, but I'm not sure why. Also for some reasons it happens only for KernelFusion tests (only they use SPIR-V 1.4 functionality?). Need to have extra investigation. Converted to draft for now.

**Comment by MrSidims** [comment]
It seems like that before going to the device's driver, the code is being but to the translator one more (or two more?) time(s). And when it's done, `--spirv-max-version` is set to 1.2, while the module can already contain SPIR-V 1.4 features. @sommerlukas @Naghasan is my understanding correct? If so, can we change `--spirv-max-version` to 1.4 in whatever tool, that fusion uses? TBH I can't find in intel/llvm, where it's being set with an exception of clang driver, where it's set to 1.4 (unless we compile to HIP, where SPIR-V is 1.1).

**Comment by sommerlukas** [comment]
> It seems like that before going to the device's driver, the code is being but to the translator one more (or two more?) time(s). And when it's done, `--spirv-max-version` is set to 1.2, while the module can already contain SPIR-V 1.4 features. @sommerlukas @Naghasan is my understanding correct? If so, can we change `--spirv-max-version` to 1.4 in whatever tool, that fusion uses? TBH I can't find in intel/llvm, where it's being set with an exception of clang driver, where it's set to 1.4 (unless we compile to HIP, where SPIR-V is 1.1).

Hi @MrSidims,
the SPIR-V version for translation to/from SPIR-V for fusion is set here: 
https://github.com/intel/llvm/blob/sycl/sycl-fusion/jit-compiler/lib/translation/SPIRVLLVMTranslation.cpp#L63

We initially chose 1.2 as the maximum value, as this is a safe default for OpenCL 2.2 devices. 

You can remove setting the maximum version [there](https://github.com/intel/llvm/blob/sycl/sycl-fusion/jit-compiler/lib/translation/SPIRVLLVMTranslation.cpp#L63), so kernel fusion would use the tool default.

**Comment by MrSidims** [comment]
@sommerlukas thanks! I've created https://github.com/intel/llvm/pull/8358 . One of the biggest reasons to move to 1.4 is that partial unrolling started to be supported only in this version, and we would lack performance in some workloads otherwise. Also our compiler implementation heavily relies on `UserSemantic `decoration, which is also added in 1.4. So while OpenCL 2.2 reasoning is legit, unfortunately we can't stick to it :( .

**Comment by sommerlukas** [comment]
@MrSidims: Thanks for the explanation! From our side, 1.2 was only the initial choice, it should be fine to switch to 1.4.

**Comment by MrSidims** [comment]
Alexey's requested change was to merge sycl branch into this one
With the tests passing I believe the PR is good to go @intel/llvm-gatekeepers

**Review by bader — APPROVED** [review]
